src/table/memo.rs

```diff
@@ -45,29 +57,55 @@ struct MemoEntry {
 /// This allows us to store into the memo-entry without acquiring a write-lock.
 /// However, using `ArcSwap` means we cannot use a `Arc<dyn Any>` or any other wide pointer.
 /// Therefore, we hide the type by transmuting to `DummyMemo`; but we must then be very careful
-/// when freeing `MemoEntryData` values to transmute things back. See the `Drop` impl for
-/// [`MemoEntry`][] for details.
-struct MemoEntryData {
+/// when freeing `MemoEntryData` values to transmute things back.
+#[derive(Default)]
+struct MemoEntry {
+    arc_swap: ManuallyDrop<ArcSwapOption<DummyMemo>>,
+}
+
+struct MemoEntryType {
     /// The `type_id` of the erased memo type `M`
     type_id: TypeId,
 
     /// A pointer to `std::mem::drop::<Arc<M>>` for the erased memo type `M`
     to_dyn_fn: fn(Arc<DummyMemo>) -> Arc<dyn Memo>,
+}
 
-    /// An [`ArcSwap`][] to a `Arc<M>` for the erased memo type `M`
-    arc_swap: ArcSwap<DummyMemo>,
+/// Using `RwLock<Vec>` costs too much, unfortunately, so we're using a data structure
+/// that is free for reads.
+struct MemoEntryTypeLock {
+    data: UnsafeCell<MemoEntryType>,
+    /// The key is: as long as `initialized` is `false`, you cannot make any shared reference
+    /// to `data` (because there might be writes), and every write must be guarded by `init_lock`.
+    /// Once someone writes into `data`, they change `initialized` to true, and from this point
+    /// only shared references to `data` are allowed, no mutation.
+    initialized: AtomicBool,
+    /// If `initialized` is `false`, we take this lock or block on it.
+    init_lock: Mutex<()>,
 }
 
-/// Dummy placeholder type that we use when erasing the memo type `M` in [`MemoEntryData`][].
-struct DummyMemo {}
+// SAFETY: We handle synchronization.
+unsafe impl Sync for MemoEntryTypeLock {}
 
-impl MemoTable {
-    fn to_dummy<M: Memo>(memo: Arc<M>) -> Arc<DummyMemo> {
-        unsafe { std::mem::transmute::<Arc<M>, Arc<DummyMemo>>(memo) }
-    }
+#[derive(Default)]
+pub struct MemoTableTypes {
+    types: AppendOnlyVec<MemoEntryTypeLock>,
```

Comment:
Huh. I'll have to read this again in a bit, but I'm feeling a bit confused -- why do we need a lock per entry, actually? I kind of expected the memo type information to be more centralized.

Reply:
Because if I put only one global lock, it is too slow (benchmarked on rust-analyzer).